### PR TITLE
CA-284857: VM clean shutdown failed when usb is attached to the vm.

### PR DESCRIPTION
**Pull request by taoyongd**
Connection refused error occured when vm is shutdown with usb attached,
which is caused by connecting qmp failed. This ticket is to define a new
exceptioin QMP_connection_error and catch it in vusb_unplug.

Signed-off-by: Taoyong Ding <taoyong.ding@citrix.com>

### CODE REVIEW

**Comment by mseri** [comment]
Looks good in principle but requires first some testing and discussion. Labeling as blocked for the moment

**Comment by taoyongd** [comment]
Hi @mseri, BVT (80643) and BST(80642) passed.

**Comment by taoyongd** [comment]
Also, @lindig and @robhoes, do you have any ideas about this change? Thanks.

**Comment by lindig** [comment]
This moves the unplug code before the shutdown code - which looks good to me. However, I find this code overall hard to follow and would suggest to restructure this:

```
List.concat
  [ List.map (fun vusb -> VUSB_unplug vusb.Vusb.id) (VUSB_DB.vusbs id)
  ; Opt.default [] 
      (Opt.map (fun x -> [ VM_shutdown_domain(id, PowerOff, x) ]) timeout)
  ; simplify @@ List.concat
      [ [ VM_destroy_device_model id; ]
      ; [ let vbds = VBD_DB.vbds id in
          Parallel (id
                   , Printf.sprintf "VBD.unplug vm=%s" id
                   , List.map (fun vbd->VBD_unplug (vbd.Vbd.id, true)) vbds);
        ]
      ; List.map (fun vif -> VIF_unplug (vif.Vif.id, true)) (VIF_DB.vifs id)
      ; List.map (fun pci -> PCI_unplug pci.Pci.id) (PCI_DB.pcis id)
      ; [ VM_destroy id ]
      ] 
  ]
```

This is not ideal either but would consider this an improvement.

**Comment by taoyongd** [comment]
And if we do this, I think other places like this also need to be changed.(VM_start etc.)

**Comment by mseri** [comment]
We were discussing this with @lindig and we agree that it's not worth making the change in this PR. Especially because it is on a feature branch and larger changes can affect very much ease of rebasing

**Comment by taoyongd** [comment]
Thanks. @mseri. So do we have a decision to merge this or another method to fix this now?

**Comment by mseri** [comment]
This is not yet clear to me. I think the discussion on the ticket needs to be resolved first

**Comment by robhoes** [comment]
I'm not in favour of changing the order of operations, as it increases the scope for things going wrong. When you shut your PC down, then you also don't first unplug your USB devices by hand.

Also VIF and VBD unplugs happen after the shutdown. This isn't to actually unplug anything from the guest (since the guest is gone already), but to clean up any state left in dom0, if needed. For VIF.unplug, I believe there isn't actually anything to clean up, and indeed exceptions are caught and ignored if the domain or device is gone already:

         with
         | (Does_not_exist(_,_)) ->
           debug "VM = %s; Ignoring missing domain" (id_of vif)
         | (Device_not_connected) ->
           debug "VM = %s; Ignoring missing device" (id_of vif)

USB should work the same, and I thought we added a similar exception handler a couple of months ago, for exactly this reason: 5887122c945a6546547d4b4b5808bfa1d872c9cf. Did that change?

**Comment by robhoes** [comment]
Perhaps one of the lower-level VUSB unplug functions need to translate the QMP error that means "domain is gone" to `Does_not_exist` to make this work.

**Comment by taoyongd** [comment]
Thanks. @robhoes. WIll update this PR.

**Comment by taoyongd** [comment]
BST 80930, BVT 80928. Hi @robhoes @lindig @mseri, please review this change. Thanks.

File: xc/device.ml [hunk 1]
```diff
@@ -1451,6 +1451,35 @@ module Vusb = struct
...
+         creatable types and instantiating objects from those types.
+        2. Qom types can be instantiated and configured directly from the QEMU
+         monitor or command-line (eg,-device, device_add).
+        3. Command example:
+         {"execute":"qom-list","arguments":{"path":"/machine/peripheral"}}
+         result:
+         {"return": [{"name": "usb1", "type": "child<usb-host>"}, {"name":"type", "type": "string"}}
+         The usb1 is added.
+    *)
+    debug "qom list to check if usb device is attached to vm";
```
Comment (lindig) [inline xc/device.ml]: This debug message need more context to be useful in the logs; it needs at least the domid. I would also move it further down and include the path because the qmp command might not be executed when `is_running` is false. However, `qmp_send_cmd` itself logs all commands - so this debug  message could be removed entirely as it does not add new information.

File: xc/device_common.ml [hunk 2]
```diff
@@ -38,6 +38,7 @@ exception Hotplug_script_expecting_field of device * string
 exception Unknown_device_type of string
 exception Unknown_device_protocol of string
 exception QMP_Error of int * string
+exception QMP_connection_error
```
Comment (lindig) [inline xc/device_common.ml]: Adding the connection path or domid would make the exception more useful

File: xc/device.ml [hunk 3]
```diff
@@ -1451,6 +1451,35 @@ module Vusb = struct
...
+    let is_running = Qemu.is_running ~xs domid in
+    if is_running then
```
Comment (lindig) [inline xc/device.ml]: I think this is as clear:
```
if Qemu.is_running ~xs domid then
```

File: xc/xenops_server_xen.ml [hunk 4]
```diff
@@ -2198,8 +2198,11 @@ module VUSB = struct
            let privileged = is_privileged vm in
            Device.Vusb.vusb_unplug ~xs ~privileged ~domid:frontend_domid ~id:(snd vusb.Vusb.id) ~hostbus:vusb.Vusb.hostbus ~hostport:vusb.Vusb.hostport
         ) Newest vm
-    with (Does_not_exist(_,_)) ->
+    with
+    | (Does_not_exist(_,_)) ->
       debug "VM = %s; VUSB = %s; Ignoring missing domain" vm (id_of vusb)
+    | (Device_not_connected) ->
+      debug "VM = %s; VUSB = %s; Ignoring missing device" vm (id_of vusb)
```
Comment (lindig) [inline xc/xenops_server_xen.ml]: This can't have the same log message as above because this is a different reason.

File: xc/device.ml [hunk 5]
```diff
@@ -1502,38 +1531,18 @@ module Vusb = struct
...
+           try
+             qmp_send_cmd domid Qmp.(Device_del id) |> ignore
+           with QMP_connection_error ->
+             raise Device_not_connected
       )
       (fun () ->
-         usb_reset_detach ~hostbus ~hostport ~domid ~privileged
+         let peripherals = qom_list ~xs ~domid in
+         let found = List.mem id peripherals in
+         if found then
```
Comment (lindig) [inline xc/device.ml]: I would keep it more compact. I don't think the extra names add much value to clarity and they are not reused.

```
if List.mem id (qom_list ~xs ~domid) then
```

File: xc/xenops_server_xen.ml (hunk 4 above)
Comment (taoyongd) [reply xc/xenops_server_xen.ml]: I think "Ignoring missing device" is different from above.

File: xc/device.ml (hunk 5 above)
Comment (taoyongd) [reply xc/device.ml]: Ok. Make sense

File: xc/device.ml (hunk 3 above)
Comment (taoyongd) [reply xc/device.ml]: Agree.

File: xc/device.ml (hunk 1 above)
Comment (taoyongd) [reply xc/device.ml]: Ok. Will remove that.

Comment (taoyongd) [reply xc/device.ml]: Updated.

File: xc/device_common.ml (hunk 2 above)
Comment (taoyongd) [reply xc/device_common.ml]: I add domid and error message for this. 
Updated.

File: xc/xenops_server_xen.ml (hunk 4 above)
Comment (lindig) [reply xc/xenops_server_xen.ml]: You are right - I misread this.

**Comment by taoyongd** [comment]
Hi @lindig , I have added the commit to this PR.

**Comment by taoyongd** [comment]
As this PR is covered in https://github.com/xapi-project/xenopsd/pull/472, so close this.
